### ISSUE TITLE
[BUG] `test_segmented_only_catches_404` fails

### ISSUE THREAD

**Issue by dotlambda** [pytube/pytube#1047]
**Describe the bug**
Test `test_segmented_only_catches_404` fails after updating to pytube 10.9.1. I'm testing in a sandbox without networking.

**Expected behavior**

**Output**
```
============================= test session starts ==============================
platform linux -- Python 3.8.9, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /build/source
collected 214 items

tests/test_captions.py ...........                                       [  5%]
tests/test_cipher.py ............                                        [ 10%]
tests/test_cli.py ....................................                   [ 27%]
tests/test_exceptions.py .........                                       [ 31%]
tests/test_extract.py .................                                  [ 39%]
tests/test_helpers.py ...........                                        [ 44%]
tests/test_itags.py ..                                                   [ 45%]
tests/test_main.py ......                                                [ 48%]
tests/test_metadata.py ..                                                [ 49%]
tests/test_parser.py .......                                             [ 52%]
tests/test_query.py ....................................                 [ 69%]
tests/test_request.py .....                                              [ 71%]
tests/test_streams.py .............................F                     [ 85%]
tests/contrib/test_channel.py .......                                    [ 89%]
tests/contrib/test_playlist.py .......................                   [100%]

=================================== FAILURES ===================================
_______________________ test_segmented_only_catches_404 ________________________

self = <urllib.request.HTTPSHandler object at 0x7ffff5d18d90>
http_class = <class 'http.client.HTTPSConnection'>
req = <urllib.request.Request object at 0x7ffff5d18a00>
http_conn_args = {'check_hostname': None, 'context': None}
host = 'r1---sn-p5qs7n7z.googlevideo.com'
h = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.

        http_class must implement the HTTPConnection API from http.client.
        """
        host = req.host
        if not host:
            raise URLError('no host given')

        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)

        headers = dict(req.unredirected_hdrs)
        headers.update({k: v for k, v in req.headers.items()
                        if k not in headers})

        # TODO(jhylton): Should this be redesigned to handle
        # persistent connections?

        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = {name.title(): val for name, val in headers.items()}

        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)

        try:
            try:
>               h.request(req.get_method(), req.selector, req.data, headers,
                          encode_chunked=req.has_header('Transfer-encoding'))

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:1354:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>, method = 'GET'
url = '/videoplayback?expire=1625549686&ei=FpfjYMPXJYH5hgalyqTIAw&ip=73.86.141.82&id=o-AEyhS7BF2_yHK4kU6zjt6LLSnnYA8-R9wFNB7...Cinitcwndbps&lsig=AG3C_xAwRAIgV-jluLsVPrhbkppQftkRPV2CbXtKp-uZId-R1dRZ4YwCIDFutxnrBAsmDXduFff3IPcvB-5hjwx34AbLotlRGvU5'
body = None
headers = {'Accept-Language': 'en-US,en', 'Connection': 'close', 'Host': 'r1---sn-p5qs7n7z.googlevideo.com', 'Range': 'bytes=0-9437183', ...}

    def request(self, method, url, body=None, headers={}, *,
                encode_chunked=False):
        """Send a complete request to the server."""
>       self._send_request(method, url, body, headers, encode_chunked)

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:1252:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>, method = 'GET'
url = '/videoplayback?expire=1625549686&ei=FpfjYMPXJYH5hgalyqTIAw&ip=73.86.141.82&id=o-AEyhS7BF2_yHK4kU6zjt6LLSnnYA8-R9wFNB7...Cinitcwndbps&lsig=AG3C_xAwRAIgV-jluLsVPrhbkppQftkRPV2CbXtKp-uZId-R1dRZ4YwCIDFutxnrBAsmDXduFff3IPcvB-5hjwx34AbLotlRGvU5'
body = None
headers = {'Accept-Language': 'en-US,en', 'Connection': 'close', 'Host': 'r1---sn-p5qs7n7z.googlevideo.com', 'Range': 'bytes=0-9437183', ...}
encode_chunked = False

    def _send_request(self, method, url, body, headers, encode_chunked):
        # Honor explicitly requested Host: and Accept-Encoding: headers.
        header_names = frozenset(k.lower() for k in headers)
        skips = {}
        if 'host' in header_names:
            skips['skip_host'] = 1
        if 'accept-encoding' in header_names:
            skips['skip_accept_encoding'] = 1

        self.putrequest(method, url, **skips)

        # chunked encoding will happen if HTTP/1.1 is used and either
        # the caller passes encode_chunked=True or the following
        # conditions hold:
        # 1. content-length has not been explicitly set
        # 2. the body is a file or iterable, but not a str or bytes-like
        # 3. Transfer-Encoding has NOT been explicitly set by the caller

        if 'content-length' not in header_names:
            # only chunk body if not explicitly set for backwards
            # compatibility, assuming the client code is already handling the
            # chunking
            if 'transfer-encoding' not in header_names:
                # if content-length cannot be automatically determined, fall
                # back to chunked encoding
                encode_chunked = False
                content_length = self._get_content_length(body, method)
                if content_length is None:
                    if body is not None:
                        if self.debuglevel > 0:
                            print('Unable to determine size of %r' % body)
                        encode_chunked = True
                        self.putheader('Transfer-Encoding', 'chunked')
                else:
                    self.putheader('Content-Length', str(content_length))
        else:
            encode_chunked = False

        for hdr, value in headers.items():
            self.putheader(hdr, value)
        if isinstance(body, str):
            # RFC 2616 Section 3.7.1 says that text default has a
            # default charset of iso-8859-1.
            body = _encode(body, 'body')
>       self.endheaders(body, encode_chunked=encode_chunked)

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:1298:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>
message_body = None

    def endheaders(self, message_body=None, *, encode_chunked=False):
        """Indicate that the last header line has been sent to the server.

        This method sends the request to the server.  The optional message_body
        argument can be used to pass a message body associated with the
        request.
        """
        if self.__state == _CS_REQ_STARTED:
            self.__state = _CS_REQ_SENT
        else:
            raise CannotSendHeader()
>       self._send_output(message_body, encode_chunked=encode_chunked)

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:1247:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>
message_body = None, encode_chunked = False

    def _send_output(self, message_body=None, encode_chunked=False):
        """Send the currently buffered request and clear the buffer.

        Appends an extra \\r\\n to the buffer.
        A message_body may be specified, to be appended to the request.
        """
        self._buffer.extend((b"", b""))
        msg = b"\r\n".join(self._buffer)
        del self._buffer[:]
>       self.send(msg)

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:1007:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>
data = b'GET /videoplayback?expire=1625549686&ei=FpfjYMPXJYH5hgalyqTIAw&ip=73.86.141.82&id=o-AEyhS7BF2_yHK4kU6zjt6LLSnnYA8-R9...deo.com\r\nUser-Agent: Mozilla/5.0\r\nAccept-Language: en-US,en\r\nRange: bytes=0-9437183\r\nConnection: close\r\n\r\n'

    def send(self, data):
        """Send `data' to the server.
        ``data`` can be a string object, a bytes object, an array object, a
        file-like object that supports a .read() method, or an iterable object.
        """

        if self.sock is None:
            if self.auto_open:
>               self.connect()

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:947:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>

    def connect(self):
        "Connect to a host on a given (SSL) port."

>       super().connect()

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:1414:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>

    def connect(self):
        """Connect to the host and port specified in __init__."""
>       self.sock = self._create_connection(
            (self.host,self.port), self.timeout, self.source_address)

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/http/client.py:918:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

address = ('r1---sn-p5qs7n7z.googlevideo.com', 443), timeout = None
source_address = None

    def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
                          source_address=None):
        """Connect to *address* and return the socket object.

        Convenience function.  Connect to *address* (a 2-tuple ``(host,
        port)``) and return the socket object.  Passing the optional
        *timeout* parameter will set the timeout on the socket instance
        before attempting to connect.  If no *timeout* is supplied, the
        global default timeout setting returned by :func:`getdefaulttimeout`
        is used.  If *source_address* is set it must be a tuple of (host, port)
        for the socket to bind as a source address before making the connection.
        A host of '' or port 0 tells the OS to use the default.
        """

        host, port = address
        err = None
>       for res in getaddrinfo(host, port, 0, SOCK_STREAM):

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/socket.py:787:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

host = 'r1---sn-p5qs7n7z.googlevideo.com', port = 443, family = 0
type = <SocketKind.SOCK_STREAM: 1>, proto = 0, flags = 0

    def getaddrinfo(host, port, family=0, type=0, proto=0, flags=0):
        """Resolve host and port into list of address info entries.

        Translate the host/port argument into a sequence of 5-tuples that contain
        all the necessary arguments for creating a socket connected to that service.
        host is a domain name, a string representation of an IPv4/v6 address or
        None. port is a string service name such as 'http', a numeric port number or
        None. By passing None as the value of host and port, you can pass NULL to
        the underlying C API.

        The family, type and proto arguments can be optionally specified in order to
        narrow the list of addresses returned. Passing zero as a value for each of
        these arguments selects the full range of results.
        """
        # We override this function since we want to translate the numeric family
        # and socket type values to enum constants.
        addrlist = []
>       for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
E       socket.gaierror: [Errno -2] Name or service not known

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/socket.py:918: gaierror

During handling of the above exception, another exception occurred:

cipher_signature = <pytube.__main__.YouTube object: videoId=2lAe1cqCOXo>

    def test_segmented_only_catches_404(cipher_signature):
        stream = cipher_signature.streams.filter(adaptive=True)[0]
        with mock.patch('pytube.request.head') as mock_head:
            mock_head.side_effect = HTTPError('', 403, 'Forbidden', '', '')
            with mock.patch("pytube.streams.open", mock.mock_open(), create=True):
                with pytest.raises(HTTPError):
>                   stream.download()

tests/test_streams.py:389:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pytube/streams.py:258: in download
    for chunk in request.stream(
pytube/request.py:157: in stream
    response = _execute_request(
pytube/request.py:37: in _execute_request
    return urlopen(request, timeout=timeout)  # nosec
/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:222: in urlopen
    return opener.open(url, data, timeout)
/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:525: in open
    response = self._open(req, data)
/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:542: in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:502: in _call_chain
    result = func(*args)
/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:1397: in https_open
    return self.do_open(http.client.HTTPSConnection, req,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <urllib.request.HTTPSHandler object at 0x7ffff5d18d90>
http_class = <class 'http.client.HTTPSConnection'>
req = <urllib.request.Request object at 0x7ffff5d18a00>
http_conn_args = {'check_hostname': None, 'context': None}
host = 'r1---sn-p5qs7n7z.googlevideo.com'
h = <http.client.HTTPSConnection object at 0x7ffff5d18dc0>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.

        http_class must implement the HTTPConnection API from http.client.
        """
        host = req.host
        if not host:
            raise URLError('no host given')

        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)

        headers = dict(req.unredirected_hdrs)
        headers.update({k: v for k, v in req.headers.items()
                        if k not in headers})

        # TODO(jhylton): Should this be redesigned to handle
        # persistent connections?

        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = {name.title(): val for name, val in headers.items()}

        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)

        try:
            try:
                h.request(req.get_method(), req.selector, req.data, headers,
                          encode_chunked=req.has_header('Transfer-encoding'))
            except OSError as err: # timeout error
>               raise URLError(err)
E               urllib.error.URLError: <urlopen error [Errno -2] Name or service not known>

/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/urllib/request.py:1357: URLError
=============================== warnings summary ===============================
tests/test_streams.py::test_segmented_only_catches_404
  /nix/store/c2rf6xcz15n1l97l5h6w7pqw1b1vxd2k-python3.8-pytest-6.2.4/lib/python3.8/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <function _TemporaryFileCloser.__del__ at 0x7ffff6de1a60>

  Traceback (most recent call last):
    File "/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/os.py", line 223, in makedirs
      mkdir(name, mode)
  FileExistsError: [Errno 17] File exists: '/build/source'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/tempfile.py", line 440, in __del__
      self.close()
    File "/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/tempfile.py", line 433, in close
      self.file.close()
  AttributeError: 'str' object has no attribute 'close'

    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

tests/contrib/test_channel.py::test_channel_video_list
  /nix/store/c2rf6xcz15n1l97l5h6w7pqw1b1vxd2k-python3.8-pytest-6.2.4/lib/python3.8/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <function _TemporaryFileCloser.__del__ at 0x7ffff6de1a60>

  Traceback (most recent call last):
    File "/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/tempfile.py", line 440, in __del__
      self.close()
    File "/nix/store/6cfajs6lsy9b4wxp3jvyyl1g5x2pjmpr-python3-3.8.9/lib/python3.8/tempfile.py", line 433, in close
      self.file.close()
  AttributeError: 'str' object has no attribute 'close'

    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED tests/test_streams.py::test_segmented_only_catches_404 - urllib.error....
============= 1 failed, 213 passed, 2 warnings in 98.86s (0:01:38) =============
```

**System information**
Please provide the following information:
- Python version: 3.8.9
- Pytube version: 10.9.1
- Command used to install pytube

**Comment by tfdahlin** [comment]
Good catch. Because I changed how stream filesize is fetched to make it faster, it means that a `HEAD` request is no longer being made on the test, and so I need to mock `request.stream` instead. I can fix this pretty quickly I think

**Comment by tfdahlin** [comment]
Should be fixed in v10.9.2, which will publish to pypi in a couple minutes
